tensorflow_addons/custom_ops/image/python/dense_image_warp_test.py

```diff
@@ -0,0 +1,233 @@
+            shape, image_type, flow_type)
+
+        interp = dense_image_warp_ops.dense_image_warp(
+            image=tf.convert_to_tensor(rand_image),
+            flow=tf.convert_to_tensor(rand_flows))
+
+        for _ in range(num_probes):
+            batch_index = np.random.randint(0, shape[0])
+            y_index = np.random.randint(0, shape[1])
+            x_index = np.random.randint(0, shape[2])
+
+            self._assert_correct_interpolation_value(
+                rand_image,
+                rand_flows,
+                interp,
+                batch_index,
+                y_index,
+                x_index,
+                low_precision=low_precision)
+
+    @tf_test_util.run_in_graph_and_eager_modes
+    def test_interpolation(self):
+        """Apply _check_interpolation_correctness() for a few sizes and
+        types."""
+
+        shapes_to_try = [[3, 4, 5, 6], [1, 5, 5, 3], [1, 2, 2, 1]]
+        for im_type in ["float32", "float64", "float16"]:
+            for flow_type in ["float32", "float64", "float16"]:
+                for shape in shapes_to_try:
+                    self._check_interpolation_correctness(
+                        shape, im_type, flow_type)
+
+    # TODO: switch to TF2 later.
+    @tf_test_util.run_deprecated_v1
+    def test_gradients_exist(self):
+        """Check that backprop can run.
+
+        The correctness of the gradients is assumed, since the forward
+        propagation is tested to be correct and we only use built-in tf
+        ops. However, we perform a simple test to make sure that
+        backprop can actually run. We treat the flows as a tf.Variable
+        and optimize them to minimize the difference between the
+        interpolated image and the input image.
+        """
+
+        batch_size, height, width, num_channels = [4, 5, 6, 7]
+        image_shape = [batch_size, height, width, num_channels]
+        image = tf.random.normal(image_shape)
+        flow_shape = [batch_size, height, width, 2]
+        init_flows = np.float32(np.random.normal(size=flow_shape) * 0.25)
+        flows = tf.Variable(init_flows)
+
+        interp = dense_image_warp_ops.dense_image_warp(image, flows)
+        loss = tf.math.reduce_mean(tf.math.square(interp - image))
+
+        optimizer = tf.optimizers.Adam(1.0)
+        grad = tf.gradients(loss, [flows])
+        opt_func = optimizer.apply_gradients(zip(grad, [flows]))
+        init_op = tf.compat.v1.global_variables_initializer()
+
+        with self.cached_session() as sess:
+            sess.run(init_op)
+            for _ in range(10):
+                sess.run(opt_func)
+
+    # TODO: run in both graph and eager modes
+    def test_size_exception(self):
+        """Make sure it throws an exception for images that are too small."""
+
+        shape = [1, 2, 1, 1]
+        msg = "Should have raised an exception for invalid image size"
+        with self.assertRaises(tf.errors.InvalidArgumentError, msg=msg):
```

Comment:
nit: self.assertRaisesRegexp with the expected error message is useful to ensure you are actually raising the error you expect, rather than catching another error and passing the text accidentally.

Reply:
+1, thank you for catching that, Karmel :-)

Reply:
Done!